File: src/main/python/AoC2024_17.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 17
#

import sys
from collections import deque

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.common import log
from aoc.vm import Instruction
from aoc.vm import Program
from aoc.vm import VirtualMachine

Input = tuple[int, int, int, list[int]]
Output1 = str
Output2 = int


TEST1 = """\
Register A: 729
Register B: 0
Register C: 0

Program: 0,1,5,4,3,0
"""
TEST2 = """\
Register A: 2024
Register B: 0
Register C: 0

Program: 0,3,5,4,3,0
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        lines = list(input_data)
        a, b, c = map(int, (lines[i][12:] for i in range(3)))
        ops = list(map(int, lines[4][9:].split(",")))
        return a, b, c, ops

    def create_instructions(self, ops: list[int]) -> list[Instruction]:
        def combo(operand: int) -> str:
            match operand:
                case 0 | 1 | 2 | 3:
                    return str(operand)
                case 4 | 5 | 6:
                    return ["*A", "*B", "*C"][operand - 4]
                case _:
                    raise ValueError

        ins = []
        ip_map = dict[int, int]()
        ip = 0
        for i in range(0, len(ops), 2):
            ip_map[i] = ip
            opcode, operand = ops[i], ops[i + 1]
            match opcode:
                case 0:
                    ins.append(Instruction.RSH("A", combo(operand)))
                    ip += 1
                case 1:
                    ins.append(Instruction.XOR("B", str(operand)))
                    ip += 1
                case 2:
                    ins.append(Instruction.SET("B", str(combo(operand))))
                    ins.append(Instruction.AND("B", "7"))
                    ip += 2
                case 3:
                    ins.append(
                        Instruction.JN0("*A", "!" + str(ip_map[operand]))
                    )
                    ip += 1
                case 4:
                    ins.append(Instruction.XOR("B", "*C"))
                    ip += 1
                case 5:
                    ins.append(Instruction.SET("X", str(combo(operand))))
                    ins.append(Instruction.AND("X", "7"))
                    ins.append(Instruction.OUT("*X"))
                    ip += 3
                case 6:
                    ins.append(Instruction.SET("C", "*B"))
                    ins.append(Instruction.RSH("C", combo(operand)))
                    ip += 2
                case 7:
                    ins.append(Instruction.SET("C", "*A"))
                    ins.append(Instruction.RSH("C", combo(operand)))
                    ip += 2
                case _:
                    raise ValueError
        return ins

    def run_program(
        self, ins: list[Instruction], a: int, b: int, c: int
    ) -> list[str]:
        output = []
        program = Program(ins, output_consumer=lambda s: output.append(s))
        program.registers["A"] = int(a)
        program.registers["B"] = int(b)
        program.registers["C"] = int(c)
        VirtualMachine().run_program(program)
        return output

    def part_1(self, input: Input) -> Output1:
        a, b, c, ops = input
        ins = self.create_instructions(ops)
        return ",".join(self.run_program(ins, a, b, c))

    def part_2(self, input: Input) -> Output2:
        _, b, c, ops = input
        ins = self.create_instructions(ops)
        wanted = list(str(_) for _ in ops)
        log(f"{wanted=}")
        seen = set([0])
        q = deque([0])
        while q:
            cand_a = q.popleft() * 8
            for i in range(8):
                na = cand_a + i
                res = self.run_program(ins, na, b, c)
                if res == wanted:
                    return na
                if res == wanted[-len(res) :] and na not in seen:  # noqa E203
                    seen.add(na)
                    log(na)
                    q.append(na)
        raise RuntimeError("unsolvable")

    @aoc_samples(
        (
            ("part_1", TEST1, "4,6,3,5,6,3,5,2,1,0"),
            ("part_2", TEST2, 117440),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 17)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_17.java
import static java.util.stream.Collectors.joining;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;
import com.github.pareronia.aoc.vm.Instruction;
import com.github.pareronia.aoc.vm.VirtualMachine;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.IntStream;

public final class AoC2024_17 extends SolutionBase<AoC2024_17.Input, String, Long> {

    private AoC2024_17(final boolean debug) {
        super(debug);
    }

    public static AoC2024_17 create() {
        return new AoC2024_17(false);
    }

    public static AoC2024_17 createDebug() {
        return new AoC2024_17(true);
    }

    @Override
    protected Input parseInput(final List<String> inputs) {
        return Input.fromInput(inputs);
    }

    @Override
    public String solvePart1(final Input input) {
        return Program.create(input.program).run(input.a, input.b, input.c).stream()
                .map(String::valueOf)
                .collect(joining(","));
    }

    @Override
    public Long solvePart2(final Input input) {
        final Program program = Program.create(input.program);
        final List<Long> ops = input.program.stream().map(Long::valueOf).toList();
        final Set<Long> seen = new HashSet<>(List.of(0L));
        final Deque<Long> q = new ArrayDeque<>(List.of(0L));
        while (!q.isEmpty()) {
            final long canda = q.pollFirst() * 8;
            for (int i = 0; i < 8; i++) {
                final long na = canda + i;
                final List<Long> res = program.run(na, input.b, input.c);
                if (res.equals(ops)) {
                    return na;
                }
                final List<Long> subList = ops.subList(ops.size() - res.size(), ops.size());
                if (res.equals(subList) && !seen.contains(na)) {
                    seen.add(na);
                    q.addLast(na);
                }
            }
        }
        throw new IllegalStateException("Unsolvable");
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "4,6,3,5,6,3,5,2,1,0"),
        @Sample(method = "part2", input = TEST2, expected = "117440")
    })
    public void samples() {}

    public static void main(final String[] args) throws Exception {
        AoC2024_17.create().run();
    }

    private static final String TEST1 =
            """
            Register A: 729
            Register B: 0
            Register C: 0

            Program: 0,1,5,4,3,0
            """;
    private static final String TEST2 =
            """
            Register A: 2024
            Register B: 0
            Register C: 0

            Program: 0,3,5,4,3,0
            """;

    record Input(long a, long b, long c, List<Integer> program) {

        public static Input fromInput(final List<String> inputs) {
            final long[] abc =
                    IntStream.range(0, 3)
                            .mapToLong(i -> Long.parseLong(inputs.get(i).substring(12)))
                            .toArray();
            final List<Integer> ops =
                    Arrays.stream(inputs.get(4).substring(9).split(","))
                            .map(Integer::valueOf)
                            .toList();
            return new Input(abc[0], abc[1], abc[2], ops);
        }
    }

    record Program(List<Instruction> instructions) {

        private static String combo(final int operand) {
            return switch (operand) {
                case 0, 1, 2, 3 -> String.valueOf(operand);
                case 4, 5, 6 -> List.of("*A", "*B", "*C").get(operand - 4);
                default -> throw new IllegalArgumentException("Unexpected value: " + operand);
            };
        }

        public static Program create(final List<Integer> ops) {
            final List<Instruction> ins = new ArrayList<>();
            int ip = 0;
            final Map<Integer, Integer> ipMap = new HashMap<>();
            for (int i = 0; i < ops.size(); i += 2) {
                final int opcode = ops.get(i);
                final int operand = ops.get(i + 1);
                ipMap.put(i, ip);
                switch (opcode) {
                    case 0:
                        {
                            ins.add(Instruction.RSH("A", combo(operand)));
                            ip++;
                            continue;
                        }
                    case 1:
                        {
                            ins.add(Instruction.XOR("B", String.valueOf(operand)));
                            ip++;
                            continue;
                        }
                    case 2:
                        {
                            ins.add(Instruction.SET("B", combo(operand)));
                            ins.add(Instruction.AND("B", "7"));
                            ip += 2;
                            continue;
                        }
                    case 3:
                        {
                            ins.add(
                                    Instruction.JN0(
                                            "*A", "!" + String.valueOf(ipMap.get(operand))));
                            ip++;
                            continue;
                        }
                    case 4:
                        {
                            ins.add(Instruction.XOR("B", "*C"));
                            ip++;
                            continue;
                        }
                    case 5:
                        {
                            ins.add(Instruction.SET("X", combo(operand)));
                            ins.add(Instruction.AND("X", "7"));
                            ins.add(Instruction.OUT("*X"));
                            ip += 3;
                            continue;
                        }
                    case 6:
                        {
                            ins.add(Instruction.SET("C", "*B"));
                            ins.add(Instruction.RSH("C", combo(operand)));
                            ip += 2;
                            continue;
                        }
                    case 7:
                        {
                            ins.add(Instruction.SET("C", "*A"));
                            ins.add(Instruction.RSH("C", combo(operand)));
                            ip += 2;
                            continue;
                        }
                    default:
                        throw new IllegalArgumentException("Unexpected value: " + opcode);
                }
            }
            return new Program(ins);
        }

        public List<Long> run(final long a, final long b, final long c) {
            final List<Long> output = new ArrayList<>();
            final com.github.pareronia.aoc.vm.Program program =
                    new com.github.pareronia.aoc.vm.Program(instructions, output::add);
            program.setRegisterValue("A", a);
            program.setRegisterValue("B", b);
            program.setRegisterValue("C", c);
            new VirtualMachine().runProgram(program);
            return output;
        }
    }
}
